cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -184,18 +280,51 @@ class InputState {
     return queue_.UnsyncFront();
   }
 
+#define LATEST_VAL_CASE(id, val)                            \
+  case Type::id: {                                          \
+    using T = typename TypeIdTraits<Type::id>::Type;        \
+    using CType = typename TypeTraits<T>::CType;            \
+    return val(data->GetValues<CType>(1)[latest_ref_row_]); \
+  }
+
   KeyType GetLatestKey() const {
-    return queue_.UnsyncFront()
-        ->column_data(key_col_index_)
-        ->GetValues<KeyType>(1)[latest_ref_row_];
+    if (key_hasher_ != NULLPTR) {
+      return key_hasher_->HashesFor(queue_.UnsyncFront().get())[latest_ref_row_];
```

Comment:
I feel I am misreading this but it looks like this is hashing the batch for every `GetLatestKey` call?

Reply:
The [`HashesFor` method caches by the batch pointer](https://github.com/apache/arrow/pull/13880/commits/e8177a3775b579cecdc74c6e4ee0e24de84ff9bf#diff-5789a42aebc3bd0f5a6db687c78ab0c3eef3e316982f99fe9e0ea21fabed354cR199-R201), so a rehash will only occur when the batch changes. This applies in the context of a particular input table, which is associated with a one key hasher (when one is defined).